### PR TITLE
15.3 rollup

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Various fixes we want to get into Update 3 for VS 2017

### CODE REVIEW

**Comment by paulvanbrenk** [comment]
@digeff can you give this a spin with PZ tomorrow, ping me on what to test.

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 1]
```diff
@@ -88,9 +88,9 @@ private void Run()
                 // The Take method will block the worker thread when there are no items left in the queue
                 // and the thread will be signalled when new items are items to the queue, or the queue is
                 // marked completed.
-                this.currentCommand = this.commandQueue.Take();
-                if (this.currentCommand != null)
+                if (this.commandQueue.TryTake(out var command, Timeout.Infinite) && command != null)
```
Comment (billti) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: What's the reason for this change? I don't see any change in semantics.

File: Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs [hunk 2]
```diff
@@ -58,6 +77,61 @@ public void SetupDebugTargetInfo(ref VsDebugTargetInfo vsDebugTargetInfo, DebugL
...
+
+            // The Node2Adapter depends on features only in Node v6+, so the old v5.4 version of node will not suffice for this scenario
+            // This node.exe will be the one used by the node2 debug adapter, not the one used to host the user code.
+            var pathToNodeExe = Path.Combine(setupInstance.GetInstallationPath(), "JavaScript\\Node.JS\\v6.4.0_x86\\Node.exe");
+
+            // We check the registry to see if any parameters for the node.exe invocation have been specified (like "--inspect"), and append them if we find them.
+            var nodeParams = NodejsProjectLauncher.CheckForRegistrySpecifiedNodeParams();
+            if (!string.IsNullOrEmpty(nodeParams))
+            {
+                pathToNodeExe = pathToNodeExe + " " + nodeParams;
```
Comment (billti) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs]: Assuming `setupInstance.GetInstallationPath()` can return a path with spaces, and this is all just being concated with args into one long string, does this require any quoting?

File: Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs [hunk 3]
```diff
@@ -58,6 +77,61 @@ public void SetupDebugTargetInfo(ref VsDebugTargetInfo vsDebugTargetInfo, DebugL
...
+
+            string trimmedPathToNode2DebugAdapter = pathToNode2DebugAdapterRuntime.Replace("\"", "");
+            if (!File.Exists(trimmedPathToNode2DebugAdapter))
+            {
+                pathToNode2DebugAdapterRuntime = Environment.ExpandEnvironmentVariables(@"""%ALLUSERSPROFILE%\" +
+                    $@"Microsoft\VisualStudio\NodeAdapter\{visualStudioInstallationInstanceID}\out\src\nodeDebug.js""");
+            }
+
+            var target = vsDebugTargetInfo.bstrExe;
+            var cwd = Path.GetDirectoryName(target); // Current working directory
```
Comment (billti) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs]: Can the user not specify the working directory to use in the NTVS property pages for the project?

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 1 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: `Take` throws an exception when the collection is marked as completed, where `TryTake` returns false. There was a racecondition I saw a couple of times between `Dispose` and `Take` being called.

File: Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs (hunk 3 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs]: Yes, this is the open folder scenario though. We don't use the project properties there.

File: Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs (hunk 2 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs]: And possibly all the other paths we handle here....

@digeff Thoughts?
